src/java/org/apache/cassandra/index/sai/utils/RangeIntersectionIterator.java

```diff
@@ -0,0 +1,229 @@
+    {
+        return new Builder(limit);
+    }
+
+    public static class Builder extends RangeIterator.Builder
+    {
+        private final int limit;
+        protected List<RangeIterator> rangeIterators = new ArrayList<>();
+
+        public Builder()
+        {
+            this(Integer.MAX_VALUE);
+        }
+
+        public Builder(int limit)
+        {
+            super(IteratorType.INTERSECTION);
+            this.limit = limit;
+        }
+
+        public RangeIterator.Builder add(RangeIterator range)
+        {
+            if (range == null)
+                return this;
+
+            if (range.getCount() > 0)
+                rangeIterators.add(range);
+            else
+                FileUtils.closeQuietly(range);
+            statistics.update(range);
+
+            return this;
+        }
+
+        public RangeIterator.Builder add(List<RangeIterator> ranges)
+        {
+            if (ranges == null || ranges.isEmpty())
+                return this;
+
+            ranges.forEach(this::add);
+            return this;
+        }
+
+        public int rangeCount()
+        {
+            return rangeIterators.size();
+        }
+
+        protected RangeIterator buildIterator()
+        {
+            rangeIterators.sort(Comparator.comparingLong(RangeIterator::getCount));
+            int initialSize = rangeIterators.size();
+            // all ranges will be included
+            if (limit >= rangeIterators.size() || limit <= 0)
+                return buildIterator(statistics, rangeIterators);
+
+            // Apply most selective iterators during intersection, because larger number of iterators will result lots of disk seek.
+            Statistics selectiveStatistics = new Statistics(IteratorType.INTERSECTION);
+            for (int i = rangeIterators.size() - 1; i >= 0 && i >= limit; i--)
+                FileUtils.closeQuietly(rangeIterators.remove(i));
+
+            for (RangeIterator iterator : rangeIterators)
+                selectiveStatistics.update(iterator);
+
+            if (Tracing.isTracing())
+                Tracing.trace("Selecting {} {} of {} out of {} indexes",
+                              rangeIterators.size(),
+                              rangeIterators.size() > 1 ? "indexes with cardinalities" : "index with cardinality",
+                              rangeIterators.stream().map(RangeIterator::getCount).map(Object::toString).collect(Collectors.joining(", ")),
+                              initialSize);
+
+            return buildIterator(selectiveStatistics, rangeIterators);
+        }
+
+        private static RangeIterator buildIterator(Statistics statistics, List<RangeIterator> ranges)
+        {
+            // if the range is disjoint or we have an intersection with an empty set,
+            // we can simply return an empty iterator, because it's not going to produce any results.
+            if (statistics.isDisjoint())
+            {
+                // release posting lists
```

Comment:
nit: Would probably remove...unless we want to comment that closing RI's possibly closes postings lists (or not) everywhere :D

Reply:
Agreed, it's not strictly true in all cases anyway.